src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1453 @@
+ *   dfs_lookup_rel("/path", "to")
+ *   dfs_lookup_rel("/path/to", "dir1")
+ *   dfs_lookup_rel("/path/to", "dir2")
+ * Coupled with a more efficient custom parse_path function,
+ * we could also reduce the intermediate allocations for each
+ * recursive subpath.
+ */
+static int
+hash_lookup(dfs_sys_t *dfs_sys, struct sys_path *sys_path)
+{
+	struct hash_hdl	*hdl;
+	d_list_t	*rlink;
+	mode_t		mode;
+	int		rc = 0;
+
+	/* If we aren't caching, just call dfs_lookup */
+	if (dfs_sys->hash == NULL) {
+		rc = dfs_lookup(dfs_sys->dfs, sys_path->dir_name, O_RDWR,
+				&sys_path->parent, &mode, NULL);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "dfs_lookup() %s failed "DF_RC"\n",
+				sys_path->dir_name, DP_RC(rc));
+			return rc;
+		}
+
+		/* We only cache directories */
+		if (!S_ISDIR(mode)) {
+			dfs_release(sys_path->parent);
+			return ENOTDIR;
+		}
+		return rc;
+	}
+
+	/* If cached, return it */
+	rlink = d_hash_rec_find(dfs_sys->hash, sys_path->dir_name,
+				sys_path->dir_name_len);
+	if (rlink != NULL) {
+		hdl = hash_hdl_obj(rlink);
+		D_GOTO(out, rc = 0);
+	}
+
+	/* Not cached, so create an entry and add it */
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL)
+		return ENOMEM;
+
+	hdl->name_len = sys_path->dir_name_len;
+	D_STRNDUP(hdl->name, sys_path->dir_name, sys_path->dir_name_len);
+	if (hdl->name == NULL)
+		D_GOTO(free_hdl, ENOMEM);
+
+	/* Start with 2 so we have exactly 1 reference left
+	 * when dfs_sys_umount is called.
+	 */
+	atomic_store_relaxed(&hdl->ref, 2);
+
+	/* Lookup name in dfs */
+	rc = dfs_lookup(dfs_sys->dfs, sys_path->dir_name, O_RDWR, &hdl->obj,
+			&mode, NULL);
+	if (rc != 0) {
+		D_ERROR("dfs_lookup() %s failed: %s\n",
+			sys_path->dir_name, strerror(rc));
+		D_GOTO(free_hdl_name, rc);
+	}
+
+	/* We only cache directories */
+	if (!S_ISDIR(mode))
+		D_GOTO(free_hdl_obj, rc = ENOTDIR);
+
+	/* call find_insert in case another thread added the same entry
+	 * after calling find.
+	 */
+	rlink = d_hash_rec_find_insert(dfs_sys->hash, hdl->name,
+				       sys_path->dir_name_len, &hdl->entry);
+	if (rlink != &hdl->entry) {
+		/* another thread beat us. Use the existing entry. */
+		sys_path->parent = hash_hdl_obj(rlink)->obj;
+		sys_path->rlink = rlink;
+		D_GOTO(free_hdl_obj, rc = 0);
+	}
+
+out:
+	sys_path->parent = hdl->obj;
+	sys_path->rlink = rlink;
+	return rc;
+
+free_hdl_obj:
+	dfs_release(hdl->obj);
+free_hdl_name:
+	D_FREE(hdl->name);
+free_hdl:
+	D_FREE(hdl);
+	return rc;
+}
+
+/**
+ * Free a struct sys_path.
+ */
+static void
+sys_path_free(dfs_sys_t *dfs_sys, struct sys_path *sys_path)
+{
+	D_FREE(sys_path->alloc);
+	if (dfs_sys->hash == NULL)
+		dfs_release(sys_path->parent);
+	else
+		d_hash_rec_decref(dfs_sys->hash, sys_path->rlink);
+}
+
+/**
+ * Set up a struct sys_path.
+ * Parse path into dirname and basename stored in a single
+ * allocation separated by null terminators.
+ * Lookup dir_name in the hash.
+ */
+static int
+sys_path_parse(dfs_sys_t *dfs_sys, struct sys_path *sys_path,
+	       const char *path)
```

Comment:
The idea behind this function is that calling the existing dirname and basename functions is inherently inefficient, because you have to strdup the path before calling each. By using custom parsing, we can reduce the 3 total allocations to only a single allocation.
Other opinions appreciated.

Reply:
it would be good here if you can split the dfs_sys part out of this, add it to an internal common header, since we could use this in other places.

Reply:
looking at this more, this is probably too specific to this use case on how we handle the constructed sting. so ignore me for now.